third_party/move/move-compiler-v2/src/pipeline/split_critical_edges_processor.rs

```diff
@@ -0,0 +1,193 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! This pass splits critical edges with empty blocks.
+//! A critical edge is an edge where the source node has multiple successors,
+//! and the target node has multiple predecessors.
+
+use move_model::{ast::TempIndex, model::FunctionEnv};
+use move_stackless_bytecode::{
+    function_target::FunctionData,
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::{AttrId, Bytecode, Label},
+};
+use std::collections::{BTreeMap, BTreeSet};
+
+pub struct SplitCriticalEdgesProcessor {}
+
+impl FunctionTargetProcessor for SplitCriticalEdgesProcessor {
+    fn process(
+        &self,
+        _targets: &mut FunctionTargetsHolder,
+        fun_env: &FunctionEnv,
+        data: FunctionData,
+        _scc_opt: Option<&[FunctionEnv]>,
+    ) -> FunctionData {
+        if fun_env.is_native() {
+            return data;
+        }
+        let mut transformer = SplitCriticalEdgesTransformation::new(data);
+        transformer.transform();
+        transformer.data
+    }
+
+    fn name(&self) -> String {
+        "SplitCriticalEdgesProcessor".to_owned()
+    }
+}
+
+struct SplitCriticalEdgesTransformation {
+    data: FunctionData,
+    // labels used in the original code and in the generated code
+    labels: BTreeSet<Label>,
+    srcs_count: BTreeMap<Label, usize>,
+}
+
+impl SplitCriticalEdgesTransformation {
+    pub fn new(data: FunctionData) -> Self {
+        let labels = Bytecode::labels(&data.code);
+        let srcs_count = count_srcs(&data.code);
+        Self {
+            data,
+            labels,
+            srcs_count,
+        }
+    }
+
+    /// Runs the transformation, which breaks critical edges with empty blocks.
+    fn transform(&mut self) {
+        let bytecodes = std::mem::take(&mut self.data.code);
+        for bytecode in bytecodes {
```

Comment:
If we refactored some of this code to take `&[Bytecode]` and return `Vec[Bytecode]`, we could add some local tests (in file unit tests) that demonstrate this pass. It is a bit unnerving to have no tests at all for this functionality.

Reply:
done